### PR TITLE
Time Travel Debugging

### PR DESCRIPTION
**Pull request by HowieG**
## 📥 Pull Request

**📘 Description**
Adding ability to cache LLM Calls with edited completions

**🔄 Related Issue (if applicable)**
Resolves ENG-305

**🧪 Testing**
tests/test_time_travel.py

### CODE REVIEW

File: tests/test_time_travel.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+try:
+    chat_completion_1 = client.chat.completions.create(
+        messages=[
+            {
+                "content": "Come up with a random superpower that isn't time travel. Just return the superpower in the format: 'Superpower: [superpower]'",
+                "role": "user",
+            }
+        ],
+        model="gpt-3.5-turbo-0125",
```
Comment (siyangqiu) [inline tests/test_time_travel.py]: This isn't really a test by pytest standards. Can we move it to the `core_manual_tests` tests folder and change the file name to not start with `test_` as tox and pytest will try to automatically run these files?

File: .gitignore [hunk 2]
```diff
@@ -161,4 +161,7 @@ cython_debug/
 
 .vscode/
 .benchmarks/
-.DS_Store
\ No newline at end of file
+.DS_Store
+
+agentops_time_travel.json
+agentops_time_travel.yaml
```
Comment (siyangqiu) [inline .gitignore]: Can use use `.agentops_time_travel.json` and `.agentops_time_travel.yaml` as the file names? The . signifies that the file should be hidden (in Unix environments), which I think should be the case here. Let me know if users should be looking at these files as then, the above comment does not apply

**Review by siyangqiu — APPROVED** [review]
Some comments to address, but 🚀🚀🚀

File: agentops/cli.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    )
+    timetravel_parser.add_argument(
+        "--off",
+        action="store_true",
+        help="Turns off Time Travel Debugging",
+    )
+
+    args = parser.parse_args()
+
+    if args.command == "timetravel":
```
Comment (braelyn-ai) [inline agentops/cli.py]: do we want to make an alias for this `tt`?

File: agentops/llm_tracker.py [hunk 4]
```diff
@@ -12,6 +12,15 @@
...
 from .helpers import check_call_stack_for_agent_id, get_ISO_time
 from .log_config import logger
+from .event import LLMEvent, ActionEvent, ToolEvent, ErrorEvent
+from .helpers import get_ISO_time, check_call_stack_for_agent_id
+import inspect
+from typing import Optional
+import pprint
+from .time_travel import (
+    fetch_completion_override_from_time_travel_cache,
+    # fetch_prompt_override_from_time_travel_cache,
```
Comment (braelyn-ai) [inline agentops/llm_tracker.py:22]: del

File: .gitignore (hunk 2 above)
Comment (HowieG) [reply .gitignore]: .agentops_time_travel.yaml can be hidden
agentops_time_travel.json isn't pretty but I think everyone said we should allow users to edit the cache offline without going through the dashboard

File: tests/test_time_travel.py (hunk 1 above)
Comment (HowieG) [reply tests/test_time_travel.py]: Yes

File: agentops/cli.py (hunk 3 above)
Comment (HowieG) [reply agentops/cli.py]: Yes

File: agentops/llm_tracker.py (hunk 4 above)
Comment (HowieG) [reply agentops/llm_tracker.py:22]: no

File: agentops/time_travel.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+                )
+        except FileNotFoundError:
+            return
+
+
+def fetch_time_travel_id(ttd_id):
+    try:
+        endpoint = environ.get("AGENTOPS_API_ENDPOINT", "https://api.agentops.ai")
+        payload = json.dumps({"ttd_id": ttd_id}).encode("utf-8")
+        ttd_res = HttpClient.post(f"{endpoint}/v2/get_ttd", payload)
```
Comment (braelyn-ai) [inline agentops/time_travel.py:37]: this post needs to be wrapped in a try/catch and handle `except ApiServerException:`. this is a recent change

File: agentops/time_travel.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+                    if item["prompt"].get("type") == "chatml"
+                    else str(item["prompt"])
+                ): item["returns"]
+                for item in ttd_res.body  # TODO: rename returns to completion_override
+            }
+        }
+        with open("agentops_time_travel.json", "w") as file:
+            json.dump(prompt_to_returns_map, file, indent=4)
+
+        set_time_travel_active_state("on")
```
Comment (braelyn-ai) [inline agentops/time_travel.py]: should be a bool not a string

File: agentops/time_travel.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+    except Exception as e:
+        manage_time_travel_state(activated=False, error=e)
```
Comment (braelyn-ai) [inline agentops/time_travel.py:60]: bool here

File: agentops/time_travel.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+def set_time_travel_active_state(active_setting):
+    config_path = "agentops_time_travel.yaml"
+    try:
+        with open(config_path, "r") as config_file:
+            config = yaml.safe_load(config_file) or {}
+    except FileNotFoundError:
+        config = {}
+
+    config["Time_Travel_Debugging_Active"] = True if active_setting == "on" else False
```
Comment (braelyn-ai) [inline agentops/time_travel.py]: use bool 🔪

File: agentops/time_travel.py [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+
+    with open(config_path, "w") as config_file:
+        try:
+            yaml.dump(config, config_file)
+        except:
+            print(
+                f"🖇 AgentOps: Unable to write to {config_path}. Time Travel not activated"
```
Comment (braelyn-ai) [inline agentops/time_travel.py:112]: is there a reason to use print here and note logger?

File: tests/test_time_travel.py (hunk 1 above)
Comment (braelyn-ai) [reply tests/test_time_travel.py]: agreed ^

**Comment by braelyn-ai** [comment]
Can we add unit tests for as much of this as possible?

File: agentops/time_travel.py (hunk 5 above)
Comment (HowieG) [reply agentops/time_travel.py:37]: Done

File: agentops/time_travel.py (hunk 6 above)
Comment (HowieG) [reply agentops/time_travel.py]: Done

File: agentops/time_travel.py (hunk 7 above)
Comment (HowieG) [reply agentops/time_travel.py:60]: Done

File: agentops/time_travel.py (hunk 8 above)
Comment (HowieG) [reply agentops/time_travel.py]: Done

**Comment by HowieG** [comment]
@bboynton97 no

File: agentops/time_travel.py (hunk 9 above)
Comment (HowieG) [reply agentops/time_travel.py:112]: This happens outside of a Client session existing so no logger exists yet

**Review by siyangqiu — APPROVED** [review]

**Pull request merged**